File: javascript/ql/test/query-tests/Security/CWE-829/insecure-download.js
const nugget = require('nugget');

function foo() {
    function downloadTools(installer) {
        nugget(installer.url, {}, () => { }) // NOT OK [INCONSISTENCY] - access path too deep
    }
    var constants = {
        buildTools: {
            installerUrl: 'http://download.microsoft.com/download/5/f/7/5f7acaeb-8363-451f-9425-68a90f98b238/visualcppbuildtools_full.exe'
        }
    }
    function getBuildToolsInstallerPath() {
        const buildTools = constants.buildTools
        return {
            url: buildTools.installerUrl
        }
    }

    downloadTools(getBuildToolsInstallerPath())
}


const request = require('request');

function bar() {
    request('http://www.google.com', function () { });

    nugget("https://download.microsoft.com/download/5/f/7/5f7acaeb-8363-451f-9425-68a90f98b238/visualcppbuildtools_full.exe")

    nugget("http://example.org/unsafe.APK") // $ Alert
}

var cp = require("child_process")

function baz() {
    var url = "http://example.org/unsafe.APK"; // $ Source
    cp.exec("curl " + url, function () {}); // $ Alert

    cp.execFile("curl", [url], function () {}); // $ Alert

    nugget("ftp://example.org/unsafe.APK") // $ Alert
}

const fs = require("fs");
var writeFileAtomic = require("write-file-atomic");

function test() {
    nugget("http://example.org/unsafe", {target: "foo.exe"}, () => { }) // $ Alert

    nugget("http://example.org/unsafe", {target: "foo.safe"}, () => { })

    $.get("http://example.org/unsafe.unknown", function( data ) { // $ Alert
        writeFileAtomic('unsafe.exe', data, {}, function (err) {});
    });

    $.get("http://example.org/unsafe.unknown", function( data ) {
        writeFileAtomic('foo.safe', data, {}, function (err) {});
    });
}

File: ruby/ql/test/query-tests/security/cwe-829/insecure_download.rb
require "excon"

def foo
    def download_tools(installer)
        Excon.get(installer[:url]) # $ MISSING: Alert (requires hash flow)
    end

    constants = {
        build_tools: {
            installer_url: 'http://download.microsoft.com/download/5/f/7/5f7acaeb-8363-451f-9425-68a90f98b238/visualcppbuildtools_full.exe'
        }
    }
    def get_build_tools_installer_path
        build_tools = constants[:build_tools]
        { url: build_tools[:installer_url] }
    end

    download_tools get_build_tools_installer_path
end


def bar
    Excon.get('http://www.google.com') # GOOD

    Excon.get("https://download.microsoft.com/download/5/f/7/5f7acaeb-8363-451f-9425-68a90f98b238/visualcppbuildtools_full.exe") # GOOD

    Excon.get("http://example.org/unsafe.APK") # $ Alert
end

def baz
    url = "http://example.org/unsafe.APK" # $ Source

    Excon.get(url) # $ Alert
end

def test
    File.open("foo.exe").write(Excon.get("http://example.org/unsafe").body) # $ Alert

    File.open("foo.safe").write(Excon.get("http://example.org/unsafe").body) # GOOD

    File.write("foo.exe", Excon.get("http://example.org/unsafe").body) # $ Alert

    resp = Excon.get("http://example.org/unsafe.unknown") # $ Alert
    file = File.open("unsafe.exe", "w")
    file.write(resp.body)

    resp = Excon.get("http://example.org/unsafe.unknown")
    file = File.open("foo.safe", "w")
    file.write(resp.body) # GOOD
end

def sh
    script = Net::HTTP.new("http://mydownload.example.org").get("/myscript.sh").body # $ Alert
    system(script)
end
